File: 04/solution.scala
import scala.io.Source
import scala.util.matching.Regex


object Solution extends App {
  val lines = Source.fromFile("./input.txt").getLines
    .map(s => ("\\d+".r).findAllIn(s).map(_.toInt).toList)
    .toList

  println(solve(lines, predicatePart1) == 582)
  println(solve(lines, predicatePart2) == 893)

  def unpackList(nums: List[Int]): (Int, Int, Int, Int) = nums match {
    case List(a, b, c, d) => (a, b, c, d)
  }

  def predicatePart1(a: Int, b: Int, c: Int, d: Int): Boolean = (a >= c && b <= d) || (a <= c && b >= d)

  def predicatePart2(a: Int, b: Int, c: Int, d: Int): Boolean = c <= b && a <= d

  def solve(lines: List[List[Int]], predicate: ((Int, Int, Int, Int)) => Boolean): Int = {
    lines.map(unpackList).filter(predicate).size
  }
}

File: 04/solution.py
import re
from typing import List, Callable


def process_input_line(line: str) -> List[int]:
    nums = re.findall(r'\d+', line)
    return [int(n) for n in nums]


def condition_part1(line: List[int]) -> bool:
    a, b, c, d = line
    return (a <= c and b >= d) or (c <= a and d >= b)


def condition_part2(line: List[int]) -> bool:
    a, b, c, d = line
    return c <= b and a <= d


def solve_part1(lines: List[List[int]]) -> int:
    return len(list(filter(condition_part1, lines)))


def solve_part2(lines: List[List[int]]) -> int:
    return len(list(filter(condition_part2, lines)))


with open('input.txt', 'r') as file:
    lines = file.readlines()
    lines = [process_input_line(l) for l in lines]

    print(solve_part1(lines) == 582)
    print(solve_part2(lines) == 893)
